Cleanup envconfig usage

dcbaker opened the pull request:
This started as an attempt to remove `mesonlib.is_<platform>`, because those are almost always the wrong thing to use (they break cross compiling). Then I noticed a whole bunch of non-optimal code in doing platform detection, so I went to fix that, then I noticed that we're calling a bunch of interpreter methods, when we have access to the same information without calling into the Interpreter, so I fixed that. Plus there's some generic cleanups in here.

The end result is though that we call into the platform detection code a *lot* less, because we've had it cached the whole time.

tristan957 reviewed (commented):
> 5b56b192b33e3d11e829f592e43d6d39d73939ca

>> The problem is that we store references MachineInfo

Missing a `to`.

> a4d92e333a7d2921252d8552564125604337d370

Run on sentence in the commit message.

> d1b78c1499997dc98f0d89b91268711c45b7e336

Seems like this might be better in its own PR (so it doesn't sit waiting for a review). I don't see any relationship to previous commits.

Same for the following?

- `87ae144d4ebba6c39705616e93e2a3b2c0f91c01`
- `198a42319302c64f34995df7ae021dd2dedefe00`

> 3c9fadb4271741f507c45d291901d99a5bce886e

Missing a terminating backtick in summary.

tristan957 commented:
Dropping the test failure log here for easy viewing:

```
__________________ InternalTests.test_interpreter_unpicklable __________________
[gw0] linux -- Python 3.10.2 /usr/sbin/python3

self = <unittests.internaltests.InternalTests testMethod=test_interpreter_unpicklable>

    def test_interpreter_unpicklable(self) -> None:
        build = mock.Mock()
        build.environment = mock.Mock()
        build.environment.get_source_dir = mock.Mock(return_value='')
        with mock.patch('mesonbuild.envconfig.MachineInfo.redetect', mock.Mock()), \
                self.assertRaises(mesonbuild.mesonlib.MesonBugException):
>           i = mesonbuild.interpreter.Interpreter(build, mock=True)

unittests/internaltests.py:1592: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
mesonbuild/interpreter/interpreter.py:304: in __init__
    'build_machine': OBJ.MachineHolder(self.environment.machines.build, self),
mesonbuild/interpreter/interpreterobjects.py:592: in __init__
    super().__init__(machine_info, interpreter)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def __init__(self, obj: InterpreterObjectTypeVar, interpreter: 'Interpreter') -> None:
        super().__init__(subproject=interpreter.subproject)
        # This causes some type checkers to assume that obj is a base
        # HoldableObject, not the specialized type, so only do this assert in
        # non-type checking situations
        if not T.TYPE_CHECKING:
>           assert isinstance(obj, HoldableTypes), f'This is a bug: Trying to hold object of type `{type(obj).__name__}` that is not in `{HoldableTypes}`'
E           AssertionError: This is a bug: Trying to hold object of type `Mock` that is not in `(<class 'mesonbuild.utils.core.HoldableObject'>, <class 'int'>, <class 'bool'>, <class 'str'>, <class 'list'>, <class 'dict'>)`

mesonbuild/interpreterbase/baseobjects.py:146: AssertionError
```

tristan957 commented on mesonbuild/envconfig.py:
```diff
@@ -152,6 +155,180 @@
+            # Ignore compilers that do not support has_builtin_define.
+            pass
+    return False
+
```
Need an extra newline?

dcbaker commented on mesonbuild/dependencies/cuda.py:
```diff
@@ -218,8 +217,9 @@ def _strip_patch_version(cls, version: str) -> str:
     def _detect_arch_libdir(self) -> str:
-        arch = detect_cpu_family(self.env.coredata.compilers.host)
         machine = self.env.machines[self.for_machine]
+        # XXX: Why is this looking at build and not self.for_machine?
+        arch = self.env.machines.build.cpu_family
```
No, I thought that at first too, but `detect_cpu_family()` calls a bunch of `mesonlib.*` method, which all act on the python you're running, which means build machine

tristan957 replied:
Oh buddy. I do think solving the question in the comment is important.

eli-schwartz replied:
> it was the *others* that are looking at the build machine specifically.

They were?

I'm just really casually looking at this so I'm sure I can easily be missing something, but they are all using detect_cpu_family -- which performs mesonlib-based trials, i.e. build_machine ones -- and fine tune it with a *host* compiler.

I am not entirely sure what that *ever* does in a cross build, but mechanically they seem to be the same across all of these call sites.

The commit message seems wrong, anyway, they aren't (merely) silly. I'd kind of like the commit message to be expanded with an explanation of why any specific changes are either semantics-preserving or a bugfix (depending on which is true).

dcbaker replied:
I've restructured the commits a little, and moved the fixes up earlier in the series in a more verbose way, then simplified them later when we can. Does the updated version look better?

dcbaker commented on mesonbuild/interpreter/interpreter.py:
```diff
@@ -315,6 +315,8 @@ def __init__(
         self.build_holder_map()
         self.user_defined_options = user_defined_options
+
+        # Tracked compilers that are part of this subproject only.
```
Maybe? I think my reasoning for not renaming is that I've got a giant rework of the compiler going at the same time and I want to minimize my own rebase headache :)

tristan957 replied:
Fine with me!

eli-schwartz commented on mesonbuild/envconfig.py:
```diff
@@ -11,19 +11,22 @@
 import subprocess
+import sys
 import typing as T
-from enum import Enum
```
These changes seem like pointless churn, but in particular I think it makes a lot of sense to have `__future__` be its own "group" the same way stdlib, first-party, and third-party are each their own groups.

Since, the purpose of `__future__` imports is to act as a bytecode compiler cookie rather than actually importing useful objects. 

(Also you really don't want it to get jumbled into stdlib imports and then accidentally moved away from the first line.)

dcbaker replied:
I can move the `__future__` to it's own block.

eli-schwartz commented on mesonbuild/msetup.py:
```diff
@@ -219,18 +219,15 @@ def _generate(self, env: environment.Environment) -> None:
                             (mlog.log if env.is_cross_build() else mlog.debug))
-        build_machine = intr.builtin['build_machine']
-        host_machine = intr.builtin['host_machine']
-        target_machine = intr.builtin['target_machine']
```
Huh... before this PR, was the interpreter copy of each machine guaranteed to be the same one from the original env object we passed in? I *think* so but I'm not sure.

dcbaker replied:
that's.... actually a really good question

dcbaker commented on mesonbuild/envconfig.py:
```diff
@@ -355,6 +356,17 @@ def is_irix(self) -> bool:
+        return self.system == 'aix'
+
+    def is_wsl(self) -> bool:
+        return self.is_linux() and 'microsoft' in platform.release().lower()
```
Yeah.... The MachineInfo's almost need to know which machine they're pointing at...

dcbaker commented on mesonbuild/envconfig.py:
```diff
@@ -355,6 +356,17 @@ def is_irix(self) -> bool:
         """Machine is IRIX?"""
         return self.system.startswith('irix')
 
+    def is_qnx(self) -> bool:
+        """Machine is QNX?"""
+        return self.system == 'qnx'
+
+    def is_aix(self) -> bool:
+        """Machine is AIX?"""
+        return self.system == 'aix'
+
+    def is_wsl(self) -> bool:
+        return self.is_linux() and 'microsoft' in platform.release().lower()
```
My thought was that I'd really like to not have the `mesonlib.is_*` methods at all, because they often are reached for when they're incorrect because they're easy. In pretty much every case we use them except in our own test framework they're either wrong, or we have access to `environment.machines` anyway, but there's a couple of big changes required to make that possible after all of this work (the linkers and compilers need to have an `Environment` reference internally.)

dcbaker commented on mesonbuild/envconfig.py:
```diff
@@ -469,11 +471,27 @@ def from_literal(cls, literal: T.Dict[str, str]) -> 'MachineInfo':
+
+        :param compilers: A dictionary of {lang: compiler}
+        """
+        if self._allow_redetect:
```
after we've re-detected once already. Basically we do an initial "guess" even if you don't have any compilers yet, then we go back later once there's a compiler, and try to get more accurate information. We end up running that for both the host and the build in a cross setup, but I'm not convinced that we should be running that for the host compiler ever, since that's ignoring what a user put in their cross file...

dcbaker commented:
I've fixed (hopefully) a bunch of the issues you've pointed out. I've also added a WIP patch on top of this, so we'll see what it says. There's still more review feedback to look at

dcbaker commented:
That msys gccx686 failure is interesting... it's showing the build and host as x86_64... that's not right. hmmm...

eli-schwartz commented on run_single_test.py:
```diff
@@ -39,11 +40,13 @@ def main() -> None:
     parser.add_argument('--cross-file', action='store', help='File describing cross compilation environment.')
     parser.add_argument('--native-file', action='store', help='File describing native compilation environment.')
     parser.add_argument('--use-tmpdir', action='store_true', help='Use tmp directory for temporary files.')
+    parser.add_argument('--quick', action='store_true', help='Skip some compiler and tool checking')
```
I usually use `./run_project_tests.py --only common/1' '` or similar here, because I keep forgetting that run_single_test correctly handles failing/ now. Maybe I should switch over...

eli-schwartz replied:
Cherry-picked to master.
